### PR TITLE
Mechanical changes for Firewall Rules

### PR DESCRIPTION
**Pull request by SimonRichardson**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

The firewall rules importing task was wrong, most if not all the
importing logic, doesn't validate what's coming in and assumes because
the newly created model is a blank slate, that it can just insert
documents without retry logic.

When making this change to be in line with all the other tasks, I put
the firewallrules type methods back to what they where originally.

In doing these changes I added missing tests for the migration import
tasks, so we can at least verify that our import script works for both
integration and unit level testing.

## QA steps

Ensure all the tests pass.

```sh
juju bootstrap aws test
juju add-model foo
juju set-firewall-rule ssh --whitelist xxx.xxx.xxx.xxx/xxx

juju bootstrap aws other

juju switch test
juju migrate foo other

juju switch foo
```

Open a connection to mongo and ensure that the firewall rule was migrated

```sh
db.firewallRules.find().pretty()
```

Bonus points:
Expose an endpoint after deployment and look in aws console to see the firewall rule is correct opened and migrated.

### CODE REVIEW

File: state/migration_import_test.go [hunk 1]
```diff
@@ -1309,31 +1309,33 @@ func (s *MigrationImportSuite) TestSpaces(c *gc.C) {
 
 func (s *MigrationImportSuite) TestFirewallRules(c *gc.C) {
```
Comment (nammn) [inline state/migration_import_test.go]: Is the test about explicit overwriting of existing rules? 
If yes I would rename test method imo.
`TestOverwriteFirewallRules` or something along the line, wdyt?

**Comment by nammn** [comment]
```
Expose an endpoint after deployment and look in aws console to see the firewall rule is correct opened and migrated.
```

Might want to note that this only works for ssh though.

File: state/firewallrules.go [hunk 2]
```diff
@@ -28,17 +28,22 @@ import (
...
 func NewFirewallRule(serviceType firewall.WellKnownServiceType, cidrs []string) FirewallRule {
-	return FirewallRule{whitelistCIDRs: cidrs, wellKnownService: serviceType}
+	return FirewallRule{
+		id:               string(serviceType),
+		whitelistCIDRs:   cidrs,
+		wellKnownService: serviceType,
+	}
 }
 
+// ID returns the underlying Firewall ID
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/firewallrules.go]: ID is an ID, it's just coincidental that it's the same as the service type. How that's implemented under the hood, shouldn't care for the consumer of type.

File: state/migration_import_tasks.go [hunk 3]
```diff
@@ -165,29 +165,40 @@ func (s stateDocumentFactoryShim) MakeStatusOp(globalKey string, doc statusDoc)
...
+
+	ops := make([]txn.Op, len(firewallRules))
+	for i, rule := range firewallRules {
+		serviceType := rule.WellKnownService()
+		doc := firewallRulesDoc{
+			Id:               serviceType,
+			WellKnownService: serviceType,
+			WhitelistCIDRS:   rule.WhitelistCIDRs(),
+		}
+		ops[i] = txn.Op{
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/migration_import_tasks.go]: Exactly, it's important to be consistent. If we're making fundamental changes where state becomes the arbiter of changes to all migrations entities, then we should either make more changes to be consistent or add documentation around why we believe that is the right thing to do.

For example, the documentation for exporting migration changes: https://github.com/juju/juju/blob/develop/state/migration_export.go#L28-L48

File: state/firewallrules.go [hunk 4]
```diff
@@ -98,53 +145,37 @@ func (fw *firewallRulesState) Save(rule FirewallRule) error {
...
-		}
-	}
-	serviceStr := string(rule.WellKnownService())
-	doc := firewallRulesDoc{
-		Id:               serviceStr,
-		WellKnownService: serviceStr,
-		WhitelistCIDRS:   rule.WhitelistCIDRs(),
-	}
+// Remove deletes the specified firewall rule.
+func (fw *firewallRulesState) Remove(firewallRuleID firewall.WellKnownServiceType) error {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/firewallrules.go]: Good catch, I'll add tests for this one.

File: state/migration_import_test.go (hunk 1 above)
Comment (SimonRichardson) [reply state/migration_import_test.go]: When migrating from one model to another, I want to remove the old firewall rule during an export and import, to ensure that it's not the old rule that I'm picking up. So when I'm testing in the new state, it's in no way the old rule.

File: state/firewallrules.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nammn) [reply state/firewallrules.go]: I see your point, IMO that leads to comments which are just there out of consistency reasons.

File: state/migration_import_test.go (hunk 1 above)
Comment (nammn) [reply state/migration_import_test.go]: The test itself makes a lot of sense to me. I was thinking about renaming the test name as one could argue that the name is too generic.
IMO your above description would great as a method name/method description.

File: state/migration_import_tasks_test.go [hunk 5]
```diff
@@ -622,3 +622,105 @@ func (s *MigrationImportTasksSuite) externalController(ctrl *gomock.Controller,
 	entity.EXPECT().Models().Return(models)
 	return entity
 }
+
+/////
```
Comment (manadart) [inline state/migration_import_tasks_test.go]: Leftover in-progress marker?

File: state/migration_internal_test.go [hunk 6]
```diff
@@ -212,8 +214,6 @@ func (s *MigrationSuite) TestKnownCollections(c *gc.C) {
 	todoCollections := set.NewStrings(
 		// uncategorised
 		//Cross Model Relations - TODO
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/migration_internal_test.go]: I'm assuming so, I'll just assume that `dockerResourcesC` is placed under the wrong stanza.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**